ISPN-9784 Replace Distributed Executor with Cluster Executor

wburns opened the pull request:
https://issues.jboss.org/browse/ISPN-9784

wburns commented:
Testing out CI.

wburns commented:
Some CI failures.

wburns commented:
Okay got it down to 1 failure, but haven't had any luck reproducing it yet.

danberindei commented on core/src/main/java/org/infinispan/manager/DefaultCacheManager.java:
```diff
@@ -1059,10 +1059,14 @@ public Health getHealth() {
    @Override
    public ClusterExecutor executor() {
-      if (globalComponentRegistry.getStatus() != ComponentStatus.RUNNING) {
+      authzHelper.checkPermission(AuthorizationPermission.ADMIN);
```
Why not the `EXEC` permission?

wburns replied:
Whoops, yeah it should be `EXEC` :)

wburns replied:
Actually thinking about this more I can see this as `ADMIN` like I had before as well since it is at the cache manager level and many users may want to start a cache etc. I think having it more strict for now is a bit safer.

danberindei commented on core/src/main/java/org/infinispan/manager/impl/UnwrappingEmbeddedCacheManager.java:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * EmbeddedCacheManager used for cluster executor invocation so that caches are not wrapped with security
+ * since invoking via ClusterExecutor already requires ADMIN privileges
```
I'd rather require only `EXEC` for the cluster executor and pass the security principal to the remote nodes, assuming it is possible.

wburns replied:
The `DistributedExecutor` didn't currently support this, I think we need a separate JIRA to handle this though, as this will add a lot more changes to `ClusterExecutor` I wasn't planning for this PR.

danberindei replied:
Ok, please create a separate JIRA.

wburns replied:
Created https://issues.jboss.org/browse/ISPN-9914

danberindei commented on core/src/main/java/org/infinispan/notifications/cachelistener/cluster/impl/BatchingClusterEventManagerImpl.java:
```diff
@@ -85,36 +95,38 @@ public void addTargets(Address address, UUID identifier, Collection<ClusterEvent
          for (Entry<Address, TargetEvents<K, V>> entry : targets.entrySet()) {
             TargetEvents<K, V> value = entry.getValue();
             if (value.events.size() > 1) {
+               CompletableFuture<Void> future = executor.filterTargets(Collections.singleton(entry.getKey())).submitConsumer(
```
Nitpicking: I try to split chained method calls before `.` instead of after `(`, the way IntelliJ seems to like it. Same on line 115.

wburns replied:
Yeah I meant to do this. Guessing I added the filter later and didn't separate the invocations by line.

danberindei commented on core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java:
```diff
@@ -453,11 +455,13 @@ public boolean ownsData() {
-               callable.call();
+               // TODO: need security check?
+               // We have to invoke a separate method as we can't retrieve the cache as it is still starting
+               callable.accept(cache.getCacheManager(), cache);
```
TBH I always found it odd that we'd get back a `Callable` instead of a POJO with the information we need to register the listener ourselves. If we did the registration ourselves we'd just call a method on the `CacheNotifier`, we wouldn't need the cache.

wburns replied:
There are a couple parts to this reason. The first is that because this is invoked remotely it was using `DistributedExecutor` to do that instead of having to add a whole new command. This therefore requires the `Callable` approach. The second item is then the `StateConsumerImpl` doesn't need to know how to register it properly, if you notice there is quite a bit of code to make sure the listener is only registered once - which I wouldn't think the `StateConsumerImpl` would want in it. Also the `Callable` has changed a few times already - which isolates how many class changes are required.

danberindei replied:
The callable is the response from the remote node, so it could have been any serializable POJO. And `StateConsumerImpl` can delegate the registration to `ClusterCacheNotifier`, just like `StateProviderImpl` uses `retrieveClusterListenerCallablesToInstall()`.

The fact that registration changes from version to version is actually one more reason to avoid the Callables, because it's much harder to ensure that changes from one version to another are compatible for the future rolling upgrades.

wburns replied:
Well in that case the POJO or Callable would be a different class or have different members, complicating the whole thing :(

It sounds like we may want to change this as you mentioned, but probably in the scope of upgrade JIRA so it is done properly.

danberindei replied:
Agree, it's out of scope, I just added the comment because I saw the TODO

wburns commented:
I am still having troubles with reproducing this issue, however the log output is a bit interesting

```
[OK:     8, KO:     0, SKIP:     0] Test starting: org.infinispan.server.memcached.MemcachedClusteredStatsTest.testSingleConnectionPerServer
2019-01-28 18:54:49.732 INFO net.spy.memcached.MemcachedConnection:  Added {QA sa=/127.0.0.1:16411, #Rops=0, #Wops=0, #iq=0, topRop=null, topWop=null, toWrite=0, interested=0} to connect queue
2019-01-28 18:54:49.742 WARN net.spy.memcached.MemcachedConnection:  Could not redistribute to another node, retrying primary node for a.
2019-01-28 18:54:49.764 INFO net.spy.memcached.MemcachedConnection:  Shut down memcached client
2019-01-28 18:54:49.775 INFO net.spy.memcached.MemcachedConnection:  Shut down memcached client
2019-01-28 18:54:49.786 INFO net.spy.memcached.MemcachedConnection:  Added {QA sa=/127.0.0.1:16411, #Rops=0, #Wops=0, #iq=0, topRop=null, topWop=null, toWrite=0, interested=0} to connect queue
2019-01-28 18:54:49.787 WARN net.spy.memcached.MemcachedConnection:  Could not redistribute to another node, retrying primary node for a.
2019-01-28 18:54:49.794 INFO net.spy.memcached.MemcachedConnection:  Added {QA sa=/127.0.0.1:16411, #Rops=0, #Wops=0, #iq=0, topRop=null, topWop=null, toWrite=0, interested=0} to connect queue
2019-01-28 18:54:49.796 WARN net.spy.memcached.MemcachedConnection:  Could not redistribute to another node, retrying primary node for a.
2019-01-28 18:54:49.804 INFO net.spy.memcached.MemcachedConnection:  Added {QA sa=/127.0.0.1:16411, #Rops=0, #Wops=0, #iq=0, topRop=null, topWop=null, toWrite=0, interested=0} to connect queue
2019-01-28 18:54:49.805 WARN net.spy.memcached.MemcachedConnection:  Could not redistribute to another node, retrying primary node for a.
2019-01-28 18:54:49.818 INFO net.spy.memcached.MemcachedConnection:  Added {QA sa=/127.0.0.1:16411, #Rops=0, #Wops=0, #iq=0, topRop=null, topWop=null, toWrite=0, interested=0} to connect queue
2019-01-28 18:54:49.819 WARN net.spy.memcached.MemcachedConnection:  Could not redistribute to another node, retrying primary node for a.
[OK:     9, KO:     0, SKIP:     0] Test succeeded: org.infinispan.server.memcached.MemcachedReplicationTest.testReplicatedAppend
2019-01-28 18:54:49.830 INFO net.spy.memcached.MemcachedConnection:  Added {QA sa=/127.0.0.1:16411, #Rops=0, #Wops=0, #iq=0, topRop=null, topWop=null, toWrite=0, interested=0} to connect queue
[OK:     9, KO:     0, SKIP:     0] Test starting: org.infinispan.server.memcached.MemcachedReplicationTest.testReplicatedCasExists
2019-01-28 18:54:49.832 WARN net.spy.memcached.MemcachedConnection:  Could not redistribute to another node, retrying primary node for a.
[OK:     9, KO:     1, SKIP:     0] Test failed: org.infinispan.server.memcached.MemcachedClusteredStatsTest.testSingleConnectionPerServer
java.lang.AssertionError: expected [2] but found [1]
	at org.testng.Assert.fail(Assert.java:94)
	at org.testng.Assert.failNotEquals(Assert.java:496)
	at org.testng.Assert.assertEquals(Assert.java:125)
	at org.testng.Assert.assertEquals(Assert.java:167)
	at org.infinispan.server.memcached.ConnectionStatsTest.testGlobalConnections(ConnectionStatsTest.java:50)
	at org.infinispan.server.memcached.MemcachedClusteredStatsTest.testSingleConnectionPerServer(MemcachedClusteredStatsTest.java:48)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.MethodInvocationHelper$1.runTestMethod(MethodInvocationHelper.java:196)
	at org.infinispan.commons.test.TestNGLongTestsHook.run(TestNGLongTestsHook.java:24)
	at org.testng.internal.MethodInvocationHelper.invokeHookable(MethodInvocationHelper.java:208)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:635)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:816)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1124)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:774)
	at org.testng.TestRunner.run(TestRunner.java:624)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:359)
	at org.testng.SuiteRunner.access$000(SuiteRunner.java:39)
	at org.testng.SuiteRunner$SuiteWorker.run(SuiteRunner.java:393)
	at org.testng.internal.thread.ThreadUtil$2.call(ThreadUtil.java:64)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

As can be seen only connections are establish on port 16411, however the test should be establishing one on 16461 (16411 + 50) as per the test. So it seems like there may be something going on there.

Also it seems a bit odd that multiple tests are creating connections to 16411, when it seems that https://github.com/infinispan/infinispan/blob/master/server/memcached/src/test/java/org/infinispan/server/memcached/test/MemcachedTestingUtil.java#L102 should be generating unique ports. I will have to dig a bit more.

wburns commented:
Okay for whatever reason the local node doesn't have a connection any longer in trace. Need to dig further.

```
17:17:35,224 TRACE (testng-MemcachedClusteredStatsTest:[]) [AllClusterExecutor] Submitting callable to local node on executor thread! - Usually remote command thread pool
17:17:35,226 TRACE (testng-MemcachedClusteredStatsTest:[]) [AllClusterExecutor] Submitting consumer to single remote node - JGroups Address MemcachedClusteredStatsTest-NodeB-59661

17:17:35,228 TRACE (remote-thread-MemcachedClusteredStatsTest-NodeA-p2-t1:[]) [ResourceDMBean] Attribute numberOfLocalConnections has r=true,w=false,is=false and value 0
17:17:35,379 TRACE (remote-thread-MemcachedClusteredStatsTest-NodeB-p10-t2:[]) [ResourceDMBean] Attribute numberOfLocalConnections has r=true,w=false,is=false and value 1
```

wburns commented:
Adding messages to the channels being registered and unregistered is even more odd

```
[OK:     0, KO:     0, SKIP:     0] Test starting: org.infinispan.server.memcached.MemcachedClusteredStatsTest.testSingleConnectionPerServer
19:14:01,043 FATAL (Memcached-MemcachedClusteredStatsTest-ServerIO-3-1) [MemcachedDecoder] Channel registered for [id: 0x2669aeb0, L:/127.0.0.1:16211 - R:/127.0.0.1:45088]
19:14:01,043 FATAL (Memcached-MemcachedClusteredStatsTest-ServerIO-6-1) [MemcachedDecoder] Channel registered for [id: 0xb11c63c8, L:/127.0.0.1:16261 - R:/127.0.0.1:40170]
[OK:     0, KO:     1, SKIP:     0] Test failed: org.infinispan.server.memcached.MemcachedClusteredStatsTest.testSingleConnectionPerServer
java.lang.AssertionError: expected [2] but found [1]
	... snip ...
2019-01-31 19:14:01.363 INFO net.spy.memcached.MemcachedConnection:  Shut down memcached client
2019-01-31 19:14:01.416 INFO net.spy.memcached.MemcachedConnection:  Shut down memcached client
19:14:01,451 FATAL (Memcached-MemcachedClusteredStatsTest-ServerIO-3-1) [MemcachedDecoder] Channel unregistered for [id: 0x2669aeb0, L:/127.0.0.1:16211 ! R:/127.0.0.1:45088]
19:14:01,453 FATAL (Memcached-MemcachedClusteredStatsTest-ServerIO-6-1) [MemcachedDecoder] Channel unregistered for [id: 0xb11c63c8, L:/127.0.0.1:16261 ! R:/127.0.0.1:40170]

```

So in this case the connections should be registered. Have to look closer more.

wburns commented:
Only thing I can see is that it seems to be related to the connection being delayed as shown as connected. I have another log where the clients weren't even connected when the cluster executor was submitted. I think making sure the connection is good or adding an eventually may be best.

wburns commented:
Added in an eventually, will see what happens.

wburns commented:
The eventually seems to have fixed the issue. I will squash the commits down and we can see if CI passes again.

wburns commented:
Rebased and hopefully CI continues to pass.

vblagoje commented:
@wburns oh nooooo :-)

wburns commented:
CI seems to be passing now :)

danberindei commented on server/memcached/src/test/java/org/infinispan/server/memcached/MemcachedClusteredStatsTest.java:
```diff
@@ -45,8 +46,16 @@ public EmbeddedCacheManager createCacheManager(int index) {
-            mBeanServer);
+      ObjectName on = new ObjectName(String.format("%s:type=Server,name=%s,component=Transport", jmxDomain + "-0", "Memcached-" + jmxDomain));
+      // Now verify that via JMX as well, these stats are also as expected
+      eventuallyEquals(2, () -> {
```
Any idea why the delay is needed here but not on master? 

If it's not actually related to the the dist executor removal, I'd rather do it in a separate PR, because I think the test needs more cleanup:
* The test method is named "testSingleConnectionPerServer", but we expect 2 connections?
* The "JMX as well" comment is misleading, we only ever check JMX. Can we get the stats without JMX?
  * It was likely copied from `ConnectionStatsTest.testSingleLocalConnection()`, but it doesn't make sense there either
* All the methods in `ConnectionStatsTest` should be named `assertX`, not `testX`

wburns replied:
odd..it passed every time for me on IDE

My guess is the connection is established asynchronously and the test must be running a little quicker on CI.

The other option is I could actually fire an operation with the connection, which would guarantee it is connected. However these are done async as well, so I would still have to add in a wait somewhere. I figured this was just a bit easier.

danberindei replied:
@wburns do we know how spymemcached selects the server to execute an operation? If not, we can't guarantee each server is connected before asserting. Personally I'd just delete the test, because we don't actually control the number of connections, but I think I've delayed this PR enough :)

danberindei commented:
Integrated, thanks Will!
